Compile issue: ambigous partial ordering for simplify()

The following compile error occurs in the CI for Clang 16.0.6. It was also encountered for the upcoming Apple clang version 15.0.0.

`storm/src/storm/utility/constants.cpp:931:83: error: partial ordering for explicit instantiation of 'simplify' is ambiguous` (see [here](https://github.com/moves-rwth/storm/actions/runs/6021262873/job/16333845567#step:4:3434))

--- a/src/storm/solver/stateelimination/EliminatorBase.cpp
+++ b/src/storm/solver/stateelimination/EliminatorBase.cpp
@@ -21,6 +21,9 @@ EliminatorBase<ValueType, Mode>::EliminatorBase(storm::storage::FlexibleSparseMa
 
 template<typename ValueType, ScalingMode Mode>
 void EliminatorBase<ValueType, Mode>::eliminate(uint64_t row, uint64_t column, bool clearRow) {
+    using MatrixEntry = storm::storage::MatrixEntry<typename storm::storage::FlexibleSparseMatrix<ValueType>::index_type,
+                                                    typename storm::storage::FlexibleSparseMatrix<ValueType>::value_type>;
+
     // Start by finding the entry in the given column.
     bool hasEntryInColumn = false;
     ValueType columnValue = storm::utility::zero<ValueType>();
@@ -104,12 +107,8 @@ void EliminatorBase<ValueType, Mode>::eliminate(uint64_t row, uint64_t column, b
         // First, find the probability with which the predecessor can move to the current state, because
         // the forward probabilities of the state to be eliminated need to be scaled with this factor.
         FlexibleRowType& predecessorForwardTransitions = matrix.getRow(predecessor);
-        FlexibleRowIterator multiplyElement =
-            std::find_if(predecessorForwardTransitions.begin(), predecessorForwardTransitions.end(),
-                         [&](storm::storage::MatrixEntry<typename storm::storage::FlexibleSparseMatrix<ValueType>::index_type,
-                                                         typename storm::storage::FlexibleSparseMatrix<ValueType>::value_type> const& a) {
-                             return a.getColumn() == column;
-                         });
+        FlexibleRowIterator multiplyElement = std::find_if(predecessorForwardTransitions.begin(), predecessorForwardTransitions.end(),
+                                                           [&](MatrixEntry const& a) { return a.getColumn() == column; });
 
         // Make sure we have found the probability and set it to zero.
         STORM_LOG_THROW(multiplyElement != predecessorForwardTransitions.end(), storm::exceptions::InvalidStateException,
@@ -142,17 +141,13 @@ void EliminatorBase<ValueType, Mode>::eliminate(uint64_t row, uint64_t column, b
             }
 
             if (first2 == last2) {
-                std::copy_if(first1, last1, result,
-                             [&](storm::storage::MatrixEntry<typename storm::storage::FlexibleSparseMatrix<ValueType>::index_type,
-                                                             typename storm::storage::FlexibleSparseMatrix<ValueType>::value_type> const& a) {
-                                 return a.getColumn() != column;
-                             });
+                std::copy_if(first1, last1, result, [&](MatrixEntry const& a) { return a.getColumn() != column; });
                 break;
             }
             if (first2->getColumn() < first1->getColumn()) {
-                auto successorEntry = storm::utility::simplify(std::move(*first2 * multiplyFactor));
-                *result = successorEntry;
-                newBackwardEntries[successorOffsetInNewBackwardTransitions].emplace_back(predecessor, successorEntry.getValue());
+                ValueType successorValue = storm::utility::simplify<ValueType>((first2->getValue() * multiplyFactor));
+                *result = MatrixEntry(first2->getColumn(), successorValue);
+                newBackwardEntries[successorOffsetInNewBackwardTransitions].emplace_back(predecessor, successorValue);
                 ++first2;
                 ++successorOffsetInNewBackwardTransitions;
             } else if (first1->getColumn() < first2->getColumn()) {
@@ -162,8 +157,7 @@ void EliminatorBase<ValueType, Mode>::eliminate(uint64_t row, uint64_t column, b
                 ValueType sprod = multiplyFactor * first2->getValue();
                 ValueType sum = first1->getValue() + storm::utility::simplify(sprod);
                 auto probability = storm::utility::simplify(sum);
-                *result = storm::storage::MatrixEntry<typename storm::storage::FlexibleSparseMatrix<ValueType>::index_type,
-                                                      typename storm::storage::FlexibleSparseMatrix<ValueType>::value_type>(first1->getColumn(), probability);
+                *result = MatrixEntry(first1->getColumn(), probability);
                 newBackwardEntries[successorOffsetInNewBackwardTransitions].emplace_back(predecessor, probability);
                 ++first1;
                 ++first2;
@@ -172,9 +166,9 @@ void EliminatorBase<ValueType, Mode>::eliminate(uint64_t row, uint64_t column, b
         }
         for (; first2 != last2; ++first2) {
             if (first2->getColumn() != column) {
-                auto stateProbability = storm::utility::simplify(std::move(*first2 * multiplyFactor));
-                *result = stateProbability;
-                newBackwardEntries[successorOffsetInNewBackwardTransitions].emplace_back(predecessor, stateProbability.getValue());
+                ValueType probability = storm::utility::simplify<ValueType>(first2->getValue() * multiplyFactor);
+                *result = MatrixEntry(first2->getColumn(), probability);
+                newBackwardEntries[successorOffsetInNewBackwardTransitions].emplace_back(predecessor, probability);
                 ++successorOffsetInNewBackwardTransitions;
             }
         }
@@ -201,12 +195,8 @@ void EliminatorBase<ValueType, Mode>::eliminate(uint64_t row, uint64_t column, b
         // Delete the current state as a predecessor of the successor state only if we are going to remove the
         // current state's forward transitions.
         if (clearRow) {
-            FlexibleRowIterator elimIt =
-                std::find_if(successorBackwardTransitions.begin(), successorBackwardTransitions.end(),
-                             [&](storm::storage::MatrixEntry<typename storm::storage::FlexibleSparseMatrix<ValueType>::index_type,
-                                                             typename storm::storage::FlexibleSparseMatrix<ValueType>::value_type> const& a) {
-                                 return a.getColumn() == row;
-                             });
+            FlexibleRowIterator elimIt = std::find_if(successorBackwardTransitions.begin(), successorBackwardTransitions.end(),
+                                                      [&](MatrixEntry const& a) { return a.getColumn() == row; });
             STORM_LOG_ASSERT(elimIt != successorBackwardTransitions.end(),
                              "Expected a proper backward transition from " << successorEntry.getColumn() << " to " << column << ", but found none.");
             successorBackwardTransitions.erase(elimIt);
@@ -245,17 +235,9 @@ void EliminatorBase<ValueType, Mode>::eliminate(uint64_t row, uint64_t column, b
             }
         }
         if (isFilterPredecessor()) {
-            std::copy_if(first2, last2, result,
-                         [&](storm::storage::MatrixEntry<typename storm::storage::FlexibleSparseMatrix<ValueType>::index_type,
-                                                         typename storm::storage::FlexibleSparseMatrix<ValueType>::value_type> const& a) {
-                             return a.getColumn() != row && filterPredecessor(a.getColumn());
-                         });
+            std::copy_if(first2, last2, result, [&](MatrixEntry const& a) { return a.getColumn() != row && filterPredecessor(a.getColumn()); });
         } else {
-            std::copy_if(first2, last2, result,
-                         [&](storm::storage::MatrixEntry<typename storm::storage::FlexibleSparseMatrix<ValueType>::index_type,
-                                                         typename storm::storage::FlexibleSparseMatrix<ValueType>::value_type> const& a) {
-                             return a.getColumn() != row;
-                         });
+            std::copy_if(first2, last2, result, [&](MatrixEntry const& a) { return a.getColumn() != row; });
         }
         // Now move the new predecessors in place.
         successorBackwardTransitions = std::move(newPredecessors);

--- a/src/storm/utility/constants.cpp
+++ b/src/storm/utility/constants.cpp
@@ -3,7 +3,6 @@
 #include <cmath>
 #include <type_traits>
 
-#include "storm/storage/SparseMatrix.h"
 #include "storm/storage/sparse/StateType.h"
 
 #include "storm/exceptions/InvalidArgumentException.h"
@@ -135,24 +134,6 @@ ValueType simplify(ValueType value) {
     // In the general case, we don't do anything here, but merely return the value. If something else is
     // supposed to happen here, the templated function can be specialized for this particular type.
     return value;
-}
-
-template<typename IndexType, typename ValueType>
-storm::storage::MatrixEntry<IndexType, ValueType> simplify(storm::storage::MatrixEntry<IndexType, ValueType> matrixEntry) {
-    simplify(matrixEntry.getValue());
-    return matrixEntry;
-}
-
-template<typename IndexType, typename ValueType>
-storm::storage::MatrixEntry<IndexType, ValueType>& simplify(storm::storage::MatrixEntry<IndexType, ValueType>& matrixEntry) {
-    simplify(matrixEntry.getValue());
-    return matrixEntry;
-}
-
-template<typename IndexType, typename ValueType>
-storm::storage::MatrixEntry<IndexType, ValueType>&& simplify(storm::storage::MatrixEntry<IndexType, ValueType>&& matrixEntry) {
-    simplify(matrixEntry.getValue());
-    return std::move(matrixEntry);
 }
 
 template<typename ValueType>
@@ -926,12 +907,6 @@ template bool isConstant(double const& value);
 template bool isInfinity(double const& value);
 template bool isInteger(double const& number);
 template double simplify(double value);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, double> simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, double> matrixEntry);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, double>& simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, double>& matrixEntry);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, double>&& simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, double>&& matrixEntry);
 template std::pair<double, double> minmax(std::vector<double> const&);
 template double minimum(std::vector<double> const&);
 template double maximum(std::vector<double> const&);
@@ -997,12 +972,6 @@ template bool isAlmostOne(storm::ClnRationalNumber const& value);
 template storm::NumberTraits<ClnRationalNumber>::IntegerType convertNumber(storm::NumberTraits<ClnRationalNumber>::IntegerType const& number);
 template storm::ClnRationalNumber convertNumber(storm::ClnRationalNumber const& number);
 template storm::ClnRationalNumber simplify(storm::ClnRationalNumber value);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::ClnRationalNumber> simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::ClnRationalNumber> matrixEntry);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::ClnRationalNumber>& simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::ClnRationalNumber>& matrixEntry);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::ClnRationalNumber>&& simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::ClnRationalNumber>&& matrixEntry);
 template std::pair<storm::ClnRationalNumber, storm::ClnRationalNumber> minmax(std::map<uint64_t, storm::ClnRationalNumber> const&);
 template storm::ClnRationalNumber minimum(std::map<uint64_t, storm::ClnRationalNumber> const&);
 template storm::ClnRationalNumber maximum(std::map<uint64_t, storm::ClnRationalNumber> const&);
@@ -1028,12 +997,6 @@ template bool isAlmostOne(storm::GmpRationalNumber const& value);
 template storm::NumberTraits<GmpRationalNumber>::IntegerType convertNumber(storm::NumberTraits<GmpRationalNumber>::IntegerType const& number);
 template storm::GmpRationalNumber convertNumber(storm::GmpRationalNumber const& number);
 template storm::GmpRationalNumber simplify(storm::GmpRationalNumber value);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::GmpRationalNumber> simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::GmpRationalNumber> matrixEntry);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::GmpRationalNumber>& simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::GmpRationalNumber>& matrixEntry);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::GmpRationalNumber>&& simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, storm::GmpRationalNumber>&& matrixEntry);
 template storm::GmpRationalNumber minimum(std::map<uint64_t, storm::GmpRationalNumber> const&);
 template storm::GmpRationalNumber maximum(std::map<uint64_t, storm::GmpRationalNumber> const&);
 template storm::GmpRationalNumber minimum(std::vector<storm::GmpRationalNumber> const&);
@@ -1051,12 +1014,6 @@ template std::string to_string(storm::GmpRationalNumber const& value);
 // Instantiations for rational function.
 template RationalFunction one();
 template RationalFunction zero();
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, RationalFunction> simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, RationalFunction> matrixEntry);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, RationalFunction>& simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, RationalFunction>& matrixEntry);
-template storm::storage::MatrixEntry<storm::storage::sparse::state_type, RationalFunction>&& simplify(
-    storm::storage::MatrixEntry<storm::storage::sparse::state_type, RationalFunction>&& matrixEntry);
 
 // Instantiations for polynomials.
 template Polynomial one();

